lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -220,3 +694,263 @@ class ConnectorSourceNotify[In: Any val]
+
+    // create a promise for handling result
+    let promise = Promise[NotifyResult[In]]
+    promise.next[None](recover NotifyResultFulfill[In](stream_id, _session_id) end)
+
+    // send request to take ownership of this stream id
+    let request_id = ConnectorStreamNotifyId(stream_id, _session_id)
+    _listener.stream_notify(request_id, stream_id, stream_name,
+      point_of_reference, promise, source)
+
+  fun ref stream_notify_result(session_id: RoutingId, success: Bool,
+    stream: StreamTuple)
+  =>
+    if (session_id != _session_id) or (not _session_active) then
+      // This is a reply from a query that we'd sent in a prior TCP
+      // connection, or else the TCP connection is closed now,
+      // so ignore it.
+      // If the connection has been closed, any state about this query would
+      // have already been purged from any local state
+      return
+    end
+
+    // remove entry from _pending_notify set
+    _pending_notify.unset(stream.id)
+    if success then
+      // create _StreamState to place in _active_streams
+      let s = _StreamState(stream.id, stream.name, stream.last_acked,
+        stream.last_acked, _barrier_checkpoint_id)
+      _active_streams(stream.id) = s
+    end
+    // send response either way
+    send_notify_ack(success, stream.id, stream.last_acked)
+
+  fun ref send_notify_ack(success: Bool, stream_id: StreamId,
+    point_of_reference: PointOfReference)
+  =>
+    ifdef "trace" then
+      @printf[I32]("%s ::: send_notify_ack(%s, %s, %s)\n".cstring(),
+        WallClock.seconds().string().cstring(), success.string().cstring(),
+        stream_id.string().cstring(), point_of_reference.string().cstring())
+    end
+    let m = cwm.NotifyAckMsg(success, stream_id, point_of_reference)
+    _send_reply(_connector_source, m)
+
+  //////////////////
+  // Sending Replies
+  //////////////////
+  fun ref _to_error_state(source: (ConnectorSource[In] ref|None), msg: String): Bool
+  =>
+    _send_reply(source, cwm.ErrorMsg(msg))
+
+    _fsm_state = _ProtoFsmError
+    try (source as ConnectorSource[In] ref).close() else Fail() end
+    _continue_perhaps2()
+
+  fun ref _send_acks() =>
+    let new_credits = _max_credits - _credits
+    let acks = recover iso Array[(StreamId, PointOfReference)] end
+    for v in _pending_acks.values() do
+      ifdef "trace" then
+        @printf[I32]("%s ::: Acking stream_id %s to por %s\n".cstring(),
+          WallClock.seconds().string().cstring(),
+          v._1.string().cstring(),
+          v._2.string().cstring())
+      end
+      acks.push(v)
+    end
+    _pending_acks.clear()
+    // only send acks if there are new credits or new acks to send
+    if (new_credits > 0) or (acks.size() > 0) then
+      @printf[I32]("Sending acks for %s streams\n".cstring(),
+        _pending_acks.size().string().cstring())
+      _send_reply(_connector_source, cwm.AckMsg(new_credits, consume acks))
+    end
+    _credits = _credits + new_credits
```

Comment:
This will always set `_credits` to equal `_max_credits`.  Is that right?

Reply:
Yes.  But I wanted to use the same arithmetic that the client would be using, just in case I had an arithmetic bug(*), it'd be more obvious this way.

(*) Which I did, once upon a time.